providers/implementations/rands/drbg.c

```diff
@@ -922,3 +922,30 @@ int ossl_drbg_set_ctx_params(PROV_DRBG *drbg, const OSSL_PARAM params[])
         return 0;
     return 1;
 }
+
+/* Confirm digest is allowed to be used with a DRBG */
+int ossl_drbg_verify_digest(const EVP_MD *md)
+{
+#ifdef FIPS_MODULE
+    /* FIPS 140-3 IG D.R limited DRBG digests to a specific set */
+    static const char *const allowed_digests[] = {
+        "SHA1",                     /* SHA 1 allowed */
+        "SHA2-256", "SHA2-512",     /* non-truncated SHA2 allowed */
+        "SHA3-256", "SHA3-512",     /* non-truncated SHA3 allowed */
```

Comment:
IG D.R says:
> At the time this IG is published, there are no implementations of Hash_DRBG or HMAC_DRBG which use any of the SHA3 family hash functions.

So why are the SHA-3 hashes allowed here?

Reply:
It appears you  are correct.. More test changes for you pauli :)

Reply:
The documentation at doc/man7/EVP_RAND-HASH-DRBG.pod and doc/man7/EVP_RAND-HMAC-DRBG.pod would also need an update then.

For what it's worth, it's a mystery to me why NIST decided to build in this "if SHA-3 DRBGs ever become available, only SHA3-256 and SHA3-512 can be used" clause. It doesn't make sense from an efficiency standpoint, considering that SHA-3 is fundamentally structurally different from SHA-2. There's no truncation involved when computing SHA3-224 or SHA3-384. Consistency with SHA-2, perhaps.

Reply:
NIST is wrong :)  OpenSSL 3.0.0 FIPS provider is approved to run its DRBGs with SHA3 based digests (224, 256, 384 and 512).  Q.v. security policy page 23 in the FIPS approved Algorithms table (table 5 in the original, table 6 in the recently updated version).

This IG entire is a mystery to me, who cares if I choose to use a truncated digest for my DRBG.  Performance isn't an excuse I've seen NIST use previously, the opposite is generally the outcome of FIPS mandates.

Reply:
> The documentation at doc/man7/EVP_RAND-HASH-DRBG.pod and doc/man7/EVP_RAND-HMAC-DRBG.pod would also need an update then.

What were you thinking needed updating?

Reply:
> > The documentation at doc/man7/EVP_RAND-HASH-DRBG.pod and doc/man7/EVP_RAND-HMAC-DRBG.pod would also need an update then.
> 
> What were you thinking needed updating?

If SHA-3 was removed from this table, the documentation there would need to be changed as well. But seeing as you're not planning to remove SHA-3 here, it can stay the way it is currently.

Reply:
@t-j-h - Not sure we should be overriding the IG?

Reply:
@slontis, suggest you read the entirety of that note in the IG, rather than the clipped sample:

>At the time this IG is published, there are no implementations of Hash_DRBG or HMAC_DRBG which
use any of the SHA3 family hash functions. **If such implementations appear in the future, they shall be
using SHA3-256 or SHA3-512, and not SHA3-224 or SHA3-384**

We're in no way going against the IG, in fact we're enforcing the use of the former two adhering to the IG.